### PR TITLE
fix for CDPD-10364

### PR DESCRIPTION
**Pull request by amitagarwal06**
CDPD-10364: Have tuned hive, yarn, mapreduce configs in DE & DE HA template as per JIRA 
[CDPD-10364](https://jira.cloudera.com/browse/CDPD-10364)

List of tuned configs are -
HIVE
-----

[CDPD-9787](https://jira.cloudera.com/browse/CDPD-9787)
hive.tez.auto.reducer.parallelism=false

[CDPD-9860](https://jira.cloudera.com/browse/CDPD-9860)
hive.server2.tez.sessions.per.default.queue=4

hive.mv.files.thread=20
hive.load.dynamic.partitions.thread=20

YARN
------
[CDPD-9916](https://jira.cloudera.com/browse/CDPD-9916)
Set this only for 7.1.0 template as 7.2.0 already has the fix

yarn.log-aggregation.IFile.remote-app-log-dir to empty

[CDPD-9874](https://jira.cloudera.com/browse/CDPD-9874)
For below config removed yarn.scheduler.capacity.per-node-heartbeat.maximum-offswitch-assignments=4 entry from cluster template so that it goes to default value of 1

yarn.scheduler.capacity.per-node-heartbeat.maximum-offswitch-assignments=1

MapReduce
------------

Also added yarn-GATEWAY-BASE role config 
[OPSAPS-55437](https://jira.cloudera.com/browse/OPSAPS-55437)

mapreduce.map.memory.mb=4GB
mapreduce.reduce.memory.mb=4GB

Testing done:

- I created custom template from DE & DE HA templates with tuned values in mow-dev. And launched clusters couple of times. There were no issues. Noticed changes were applied.
- Ran some beeline queries, they worked fine.

Closes #CDPD-10364

### CODE REVIEW

**Comment by amitagarwal06** [comment]
@sunilgovind @PrabhuJoseph Kindly review YARN config changes in DE template for 7.1.0 & 7.2.0 and provide approval.

@harishjp Please review Hive config changes in DE template for 7.1.0 & 7.2.0 and provide approval.

File: core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp [hunk 1]
```diff
@@ -80,7 +84,7 @@
...
+                "value": "<configuration><property><name>yarn.scheduler.capacity.root.queues</name><value>default</value></property><property><name>yarn.scheduler.capacity.root.capacity</name><value>100</value></property><property><name>yarn.scheduler.capacity.root.default.capacity</name><value>100</value></property><property><name>yarn.scheduler.capacity.root.acl_submit_applications</name><value> </value></property><property><name>yarn.scheduler.capacity.root.acl_administer_queue</name><value> </value></property><property><name>yarn.scheduler.capacity.root.default.acl_submit_applications</name><value>*</value></property><property><name>yarn.scheduler.capacity.root.default.minimum-user-limit-percent</name><value>100</value></property><property><name>yarn.scheduler.capacity.maximum-am-resource-percent</name><value>0.33</value></property><property><name>yarn.scheduler.capacity.node-locality-delay</name><value>0</value></property><property><name>yarn.scheduler.capacity.schedule-asynchronously.maximum-threads</name><value>1</value></property><property><name>yarn.scheduler.capacity.schedule-asynchronously.scheduling-interval-ms</name><value>10</value></property></configuration>"
```
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:87]: Is the default value for off-switch assignments 1?

The YARN team is going to be putting in a patch which will help with this soon.

Comment (amitagarwal06) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:87]: > There's a few other configs that are in your doc / wiki.
> 
> hiveserver2_mv_files_thread
> hiveserver2_load_dynamic_partitions_thread_count
> hiveserver2_vectorized_adaptor_usage_mode=all
> 
> Do any of these need to be set?


@sidseth 
We definitely want to tune hiveserver2_mv_files_thread and hiveserver2_load_dynamic_partitions_thread_count but they require dynamic values depending upon the type of machine instance, this cannot be done through cluster template, hence omitted. More details about them are in the [wiki](https://cloudera.atlassian.net/wiki/spaces/ENG/pages/192414479/Config+tunings+for+Datahub+DE+cluster#3.-Hive-Tunings) 

hiveserver2_vectorized_adaptor_usage_mode=all, @rbalamohan said this is applicable only for Experian case, hence I did not change this. Experian can change this in their custom templates.



> Is the default value for off-switch assignments 1?
> 
> The YARN team is going to be putting in a patch which will help with this soon.

As far as I know default value of off-switch assignments is 1 thats why I removed this config from cluster template.

Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:87]: Technically, tez container size, YARN container size are also dynamic.

Are the defaults for the thread configs good enough?, or should better defaults be set up in the template.

I'm fine with the patch going in as is. If we can make defaults a little better - great. Eventually, and hopefully, soon - all of this gets removed from the template.

Comment (amitagarwal06) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:87]: @sidseth 
After discussing with Rajesh I have updated defaults for hive.mv.files.thread & hive.load.dynamic.partitions.thread to 20.

File: core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp [hunk 2]
```diff
@@ -191,6 +210,14 @@
...
             "name": "tez_container_size",
             "value": "4096"
+          },
+          {
+            "name": "tez_auto_reducer_parallelism",
+            "value": "false"
+          },
+          {
+            "name": "hive_service_config_safety_valve",
+            "value": "<property><name>hive.hook.proto.file.per.event</name><value>true</value></property><property><name>hive.server2.tez.sessions.per.default.queue</name><value>4</value></property>"
```
Comment (harishjp) [inline core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:220]: Overriding safety-valves in the blueprints is generally a hack and does not work in all cases. Please take a look at HiveKnoxConfigProvider and make sure you are not overriding: `String principal = ConfigUtils.getSafetyValveProperty("hive.server2.authentication.spnego.principal", "HTTP/_HOST@" + realm);` in your safety valve.

Comment (amitagarwal06) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:220]: @harishjp I understand overriding safety-valves in blueprints has issues. If one makes a config change for a safety-valve in the blueprint then all other configs inside safety-valve go away (which are generally set via java code).
We wanted to change only "hive.server2.tez.sessions.per.default.queue" which was not exposed directly. But when looked at "hive_service_config_safety_valve" in CM UI, found that only "hive.hook.proto.file.per.event=true" is being set. To not remove this while modifying blueprint I have added both the configs in blueprint.

In the meantime I have already filed a [request to expose ](https://jira.cloudera.com/browse/OPSAPS-55598) "hive.server2.tez.sessions.per.default.queue" directly in CM. When this is fixed we will remove this safety-valve from blueprint.

Also, I have verified from one of my cluster that we are not overriding/removing "hive.server2.authentication.spnego.principal" while launching cluster through updated blueprint. Its values before and after modification was hive.server2.authentication.spnego.principal=HTTP/_HOST@CLOUD-PE.XCU2-8Y8X.DEV.CLDR.WORK

Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp:220]: The addition if hive.hook.proto.file.per.event may not be necessary here after https://jira.cloudera.com/browse/OPSAPS-54706, which was committed to 2.20.

Comment (amitagarwal06) [reply core/src/main/resources/defaults/blueprints/cdp-data-engineering-710.bp]: Have removed hive.hook.proto.file.per.event from the 4 templates now.

**Comment by amitagarwal06** [comment]
@sidseth I have changed respective files to remove hive.hook.proto.file.per.event. Pls review and give +1.

**Comment by amitagarwal06** [comment]
@keyki Can you pls merge this pull request?

**Comment by keyki** [comment]
Merged.

**Pull request closed without merging**